src/ess_rhat.jl

```diff
@@ -233,25 +236,206 @@ The ESS and ``\\widehat{R}`` values can be computed for the following estimators
 - `StatsBase.mad`
 - `Base.Fix2(Statistics.quantile, p::Real)`
 
+## Types
+
+If no `estimator` is provided, the following `type`s of ESS estimates may be computed:
+- `:bulk`/`:rank`: mean-ESS computed on rank-normalized draws. This type diagnoses poor
+    convergence in the bulk of the distribution due to trends or different locations of the
+    chains.
+- `:tail`: minimum of the quantile-ESS for the symmetric quantiles where
+    `tail_prob=0.1` is the probability in the tails. This type diagnoses poor convergence in
+    the tails of the distribution. If this type is chosen, `kwargs` may contain a
+    `tail_prob` keyword.
+- `:basic`: basic ESS, equivalent to specifying `estimator=Statistics.mean`.
+
+While Bulk-ESS is conceptually related to basic ESS, it is well-defined even if the chains
+do not have finite variance.[^VehtariGelman2021]. For each parameter, rank-normalization
+proceeds by first ranking the inputs using "tied ranking" and then transforming the ranks to
+normal quantiles so that the result is standard normally distributed. This transform is
+monotonic.
+
 [^VehtariGelman2021]: Vehtari, A., Gelman, A., Simpson, D., Carpenter, B., & Bürkner, P. C. (2021).
     Rank-normalization, folding, and localization: An improved ``\\widehat {R}`` for
     assessing convergence of MCMC. Bayesian Analysis.
     doi: [10.1214/20-BA1221](https://doi.org/10.1214/20-BA1221)
     arXiv: [1903.08008](https://arxiv.org/abs/1903.08008)
 """
-function ess_rhat(samples::AbstractArray{<:Union{Missing,Real},3}; kwargs...)
-    return ess_rhat(Statistics.mean, samples; kwargs...)
+@constprop :aggressive function ess(
+    samples::AbstractArray{<:Union{Missing,Real},3};
+    estimator=nothing,
+    type=nothing,
+    kwargs...,
+)
+    if estimator !== nothing && type !== nothing
+        throw(ArgumentError("only one of `estimator` and `type` can be specified"))
+    elseif estimator !== nothing
+        return _ess(estimator, samples; kwargs...)
+    elseif type !== nothing
+        return _ess(_val(type), samples; kwargs...)
+    else
+        return _ess(Val(:basic), samples; kwargs...)
+    end
```

Comment:
Could we simplify the logic by just saying, if `estimator !== nothing` it is used and otherwise the `type` determines what ESS estimate is returned?
```suggestion
    if estimator !== nothing
        return _ess(estimator, samples; kwargs...)
    else
        return _ess(_val(type), samples; kwargs...)
    end
```
(Again, noting that I'm not a big fan of supporting both `Val` and `Symbol`)

Reply:
I do think it's important to throw an error if the user tries to specify both `type` and `estimator`. A bulk-median-ESS is meaningless.

Reply:
Hmm I was thinking that it would be sufficient to document that `estimator` would have higher precedence than `type` but I can see your point.

That makes me wonder though: Do we actually need two keyword arguments? It seems one would be sufficient if one must not specify both anyway. We could just interpret values of type `Symbol` (or `Val`) as the `type` and everything else as `estimator`?

Reply:
I liked the explicitness of calling it an estimator, especially for consistency with `mcse`. If we use `type` for `ess` as well, then we should also use `type` for `mcse`.

Reply:
Maybe that would be a good approach? An alternative could be to call it `method`.

Reply:
I prefer not to use `method`, as to me `method` is more about how an estimate is made than about what the estimate is (e.g. `method` currently used for autocovariance estimate method). I merged `estimator` and `type` as a new keyword `kind`. `kind` is about as semantically overloaded as `type` but avoids being potentially misunderstood as the type of the return object (also doesn't get colored strangely in syntax highlighters).